crates/subspace-service/src/dsn.rs

```diff
@@ -47,20 +55,37 @@ pub struct DsnConfig {
 
     /// Determines whether we allow keeping non-global (private, shared, loopback..) addresses in Kademlia DHT.
     pub allow_non_global_addresses_in_dht: bool,
+
+    /// System base path.
+    pub base_path: Option<PathBuf>,
 }
 
+type DsnProviderStorage<AS> =
+    NodeProviderStorage<PieceCache<AS>, Either<ParityDbProviderStorage, MemoryProviderStorage>>;
+
 pub(crate) async fn create_dsn_instance<Block, AS>(
     dsn_config: DsnConfig,
     piece_cache: PieceCache<AS>,
-) -> Result<(Node, NodeRunner<MemoryProviderStorage>), CreationError>
+) -> Result<(Node, NodeRunner<DsnProviderStorage<AS>>), CreationError>
 where
     Block: BlockT,
     AS: AuxStore + Sync + Send + 'static,
 {
     trace!("Subspace networking starting.");
 
-    // TODO: This should be a wrapper that handles locally cached pieces
-    let provider_storage = MemoryProviderStorage::new(peer_id(&dsn_config.keypair));
+    let peer_id = peer_id(&dsn_config.keypair);
+
+    let external_provider_storage = if let Some(path) = dsn_config.base_path {
+        let db_path = path.join("subspace_storage_providers_db").into_boxed_path();
+        let cache_size: NonZeroUsize = NonZeroUsize::new(MAX_PROVIDER_RECORDS_LIMIT)
+            .expect("Manual value should be greater than zero.");
+
+        Either::Left(ParityDbProviderStorage::new(&db_path, cache_size, peer_id)?)
```

Comment:
I guess the reason you didn't use `AuxStore` because it doesn't have iteration, but those are exactly the kinds of details I'd love to be provided in PR description to help with review process.

Reply:
Sent feature request for iteration support: https://github.com/paritytech/substrate/issues/13149